prov/efa: MR descriptor updates

rajachan opened the pull request:
Two commits in prep for FI_MR_HMEM.
```
commit d8c3b6c7395c310919244efda6517e791b343336 (HEAD -> mr-desc-updates, origin/recv-desc)
Author: Raghu Raja <craghun@amazon.com>
Date:   Mon Mar 9 17:35:05 2020 -0700

    prov/efa: Make rxr_inline_mr_reg() usable beyond the MR cache path

    The original rxr_inline_mr_reg() was written with the MR cache usage in
    mind and a world where the EFA provider did not require memory
    registrations, with the upper layer use pre-registered bounce buffers.
    In preparation for FI_MR_HMEM, this commit extends the function to be
    more useful beyond the inline registration. It can still walk the iov to
    register buffers inline, but will now do it only when a descriptor is
    not provided.

    Signed-off-by: Raghu Raja <craghun@amazon.com>
    Signed-off-by: Wei Zhang <wzam@amazon.com>

commit 16553728729d19bea599afd28a8deec03d6a8097
Author: Raghu Raja <craghun@amazon.com>
Date:   Mon Mar 9 17:37:21 2020 -0700

    prov/efa: Rework RX function signatures to propagate desc

    1e91d22 reworked the send-side function paths so that the registration
    descriptor did not get lost. This commit does the same for the
    receive-side, making sure an application-provided descriptor is used
    when available instead of falling back to bounce-buffers.

    Signed-off-by: Raghu Raja <craghun@amazon.com>
```

wzamazon commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -179,12 +183,10 @@ struct rxr_rx_entry *rxr_ep_alloc_unexp_rx_entry_for_tagrtm(struct rxr_ep *ep,
+	rx_entry = rxr_ep_get_rx_entry(ep, &msg, tag, ~0, ofi_op_tagged, 0);
+	if (OFI_UNLIKELY(!rx_entry))
 		FI_WARN(&rxr_prov, FI_LOG_CQ, "RX entries exhausted.\n");
-		return NULL;
```
The "return NULL" was dropped?

rajachan replied:
Gah, messed up in the rebase, will fix.

rajachan commented:
@wzamazon you might want to review the inline change again. Brought in some additional changes that made the function more general.

wzamazon commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -442,8 +442,8 @@ struct rxr_tx_entry *rxr_ep_alloc_tx_entry(struct rxr_ep *rxr_ep,
-void rxr_inline_mr_reg(struct rxr_domain *rxr_domain,
-		       struct rxr_tx_entry *tx_entry)
+void rxr_prepare_mr_send(struct rxr_domain *rxr_domain,
+			 struct rxr_tx_entry *tx_entry)
```
Is the space off here?

rwespetal commented on prov/efa/src/rxr/rxr_msg.c:
```diff
@@ -632,11 +626,7 @@ ssize_t rxr_msg_multi_recv(struct rxr_ep *rxr_ep, const struct iovec *iov,
-	rx_entry = rxr_ep_get_rx_entry(rxr_ep, iov, iov_count, tag,
-				       ignore, context,
-				       (rxr_ep->util_ep.caps &
-					FI_DIRECTED_RECV) ? addr :
```
Is it safe to drop the FI_DIRECTED_RECV checks? `rxr_match_addr` does look for `FI_ADDR_UNSPEC`.

rwespetal replied:
Got it, makes sense. Dropping the matching logic in that scenario sounds reasonable to me.

wckzhang commented on prov/efa/src/rxr/rxr.h:
```diff
@@ -357,6 +357,7 @@ struct rxr_rx_entry {
 
 	size_t iov_count;
 	struct iovec iov[RXR_IOV_LIMIT];
+	void *desc[RXR_IOV_LIMIT];
```
Can you comment this item so we know what it's for?

rajachan replied:
Will do.

wckzhang commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -72,13 +72,18 @@ struct rxr_rx_entry *rxr_ep_rx_entry_init(struct rxr_ep *ep,
+	if (msg->desc)
+		memcpy(&rx_entry->desc[0], msg->desc, sizeof(*msg->desc) * msg->iov_count);
+	else
+		memcpy(&rx_entry->desc[0], 0, sizeof(*msg->desc) * msg->iov_count);
```
Use memset instead of memcpy

rajachan replied:
Will do. That was a lazy copy-paste on my part.

wckzhang commented on prov/efa/src/rxr/rxr_msg.c:
```diff
@@ -993,40 +969,59 @@ ssize_t rxr_msg_trecv(struct fid_ep *ep_fid, void *buf, size_t len, void *desc,
 
-	ret = rxr_msg_generic_recv(ep_fid, msg->msg_iov, msg->iov_count, msg->addr,
-				   msg->tag, msg->ignore, msg->context,
+	msg.msg_iov = tagmsg->msg_iov;
```
I'm probably missing something. Why are you doing this? Are there things from tagmsg that you don't want to pass along?

rajachan replied:
The generic function serves any kind of recv, not just tagged ops, so it takes fi_msg as the argument, which is the lowest-common denominator of the support msg type structures.

wckzhang replied:
Ah alright

wckzhang commented on prov/efa/src/rxr/rxr_pkt_entry.h:
```diff
@@ -79,15 +79,15 @@ struct rxr_pkt_entry {
 #endif
 };
 
+#if defined(static_assert) && defined(__x86_64__)
```
Why is this change in this patch? Doesn't seem to do anything other than move if def randomly

rajachan replied:
Umm, must have been a rebase artifact. Will fix this up.

rajachan commented:
Addressed wilzhang@'s review comments. Rebased on master which now has the medium msg protocol merged. fi_rma_bw now passes in CI:
```
https://libfabric-ci.aws.a2z.com/job/multi-node/44/PROVIDER=efa,label=alinux/testReport/(root)/9_172_31_24_233_fabtests/fi_rma_bw__e_rdm__o_read__p__efa_/

server_cmd: /home/ec2-user/libfabric/fabtests/install/bin/fi_rma_bw -e rdm -o read -p "efa" -s fe80:0000:0000:0000:009b:1aff:fe08:bd52
server_stdout: |
bytes iters total time MB/sec usec/xfer Mxfers/sec
64 20k 1.2m 0.10s 13.10 4.89 0.20
256 20k 4.8m 0.09s 54.09 4.73 0.21
1k 20k 19m 0.12s 166.52 6.15 0.16
4k 20k 78m 0.13s 616.40 6.64 0.15
64k 2k 125m 0.08s 1598.95 40.99 0.02
1m 200 200m 0.06s 3455.23 303.48 0.00
client_cmd: /home/ec2-user/libfabric/fabtests/install/bin/fi_rma_bw -e rdm -o read -p "efa" -P 0 -s fe80:0000:0000:0000:00f7:24ff:fe38:7bac fe80:0000:0000:0000:009b:1aff:fe08:bd52
client_stdout: |
bytes iters total time MB/sec usec/xfer Mxfers/sec
64 20k 1.2m 0.10s 13.09 4.89 0.20
256 20k 4.8m 0.09s 54.06 4.74 0.21
1k 20k 19m 0.12s 173.54 5.90 0.17
4k 20k 78m 0.13s 616.01 6.65 0.15
64k 2k 125m 0.08s 1592.30 41.16 0.02
1m 200 200m 0.06s 3415.17 307.04 0.00
```

wckzhang commented:
Looks good to me

rajachan commented:
The changes here are self-contained to EFA so I can't imagine Intel CI failure being a functional issue. Going to assume it is a CI issue and merge this PR in a little while.
